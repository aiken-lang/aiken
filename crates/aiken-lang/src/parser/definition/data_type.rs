use crate::{
    ast,
    parser::{annotation, error::ParseError, token::Token, utils},
};
use chumsky::prelude::*;

pub fn parser() -> impl Parser<Token, ast::UntypedDefinition, Error = ParseError> {
    let unlabeled_constructor_type_args = annotation()
        .map_with_span(|annotation, span| ast::RecordConstructorArg {
            label: None,
            annotation,
            tipo: (),
            doc: None,
            location: span,
        })
        .separated_by(just(Token::Comma))
        .allow_trailing()
        .delimited_by(just(Token::LeftParen), just(Token::RightParen));

    let constructors = select! {Token::UpName { name } => name}
        .then(
            choice((
                labeled_constructor_type_args(),
                unlabeled_constructor_type_args,
            ))
            .or_not(),
        )
        .map_with_span(|(name, arguments), span| ast::RecordConstructor {
            location: span,
            arguments: arguments.unwrap_or_default(),
            name,
            doc: None,
            sugar: false,
        })
        .repeated()
        .delimited_by(just(Token::LeftBrace), just(Token::RightBrace));

    let record_sugar = labeled_constructor_type_args().map_with_span(|arguments, span| {
        vec![ast::RecordConstructor {
            location: span,
            arguments,
            doc: None,
            name: String::from("_replace"),
            sugar: true,
        }]
    });

    utils::optional_flag(Token::Pub)
        .then(utils::optional_flag(Token::Opaque))
        .then(utils::type_name_with_args())
        .then(choice((constructors, record_sugar)))
        .map_with_span(
            |(((public, opaque), (name, parameters)), constructors), span| {
                ast::UntypedDefinition::DataType(ast::DataType {
                    location: span,
                    constructors: if constructors.is_empty() {
                        vec![ast::RecordConstructor {
                            location: span,
                            arguments: vec![],
                            doc: None,
                            name: name.clone(),
                            sugar: true,
                        }]
                    } else {
                        constructors
                            .into_iter()
                            .map(|mut constructor| {
                                if constructor.sugar {
                                    constructor.name.clone_from(&name);
                                }

                                constructor
                            })
                            .collect()
                    },
                    doc: None,
                    name,
                    opaque,
                    parameters: parameters.unwrap_or_default(),
                    public,
                    typed_parameters: vec![],
                })
            },
        )
}

fn labeled_constructor_type_args()
-> impl Parser<Token, Vec<ast::RecordConstructorArg<()>>, Error = ParseError> {
    select! {Token::Name {name} => name}
        .then_ignore(just(Token::Colon))
        .then(annotation())
        .map_with_span(|(name, annotation), span| ast::RecordConstructorArg {
            label: Some(name),
            annotation,
            tipo: (),
            doc: None,
            location: span,
        })
        .separated_by(just(Token::Comma))
        .allow_trailing()
        .delimited_by(just(Token::LeftBrace), just(Token::RightBrace))
}

#[cfg(test)]
mod tests {
    use crate::assert_definition;

    #[test]
    fn custom_type() {
        assert_definition!(
            r#"
            type Option<a> {
              Some(a, Int)
              None
              Wow { name: Int, age: Int }
            }
            "#
        );
    }

    #[test]
    fn opaque_type() {
        assert_definition!(
            r#"
            pub opaque type User {
              name: _w
            }
            "#
        );
    }

    #[test]
    fn record_sugar() {
        assert_definition!(
            r#"
            pub type Foo {
              wow: Int,
            }
            "#
        );
    }

    #[test]
    fn empty_record_sugar() {
        assert_definition!(
            r#"
            pub type Foo {
            }
            "#
        );
    }
}
