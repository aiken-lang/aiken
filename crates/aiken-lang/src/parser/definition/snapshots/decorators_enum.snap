---
source: crates/aiken-lang/src/parser/definition/data_type.rs
description: "Code:\n\npub type L2Datum {\n  @tag(117)\n  Inactive\n\n  @tag(118)\n  BatchTxs(List<ByteArray>)\n\n  @tag(1170)\n  Committed(Bool, List<ByteArray>)\n\n  @tag(1171)\n  Withdrawn(Bool, @len(32) ByteArray)\n}\n"
---
DataType(
    DataType {
        decorators: [],
        constructors: [
            RecordConstructor {
                decorators: [
                    Tag {
                        value: "117",
                        base: Decimal {
                            numeric_underscore: false,
                        },
                    },
                ],
                location: 21..41,
                name: "Inactive",
                arguments: [],
                doc: None,
                sugar: false,
            },
            RecordConstructor {
                decorators: [
                    Tag {
                        value: "118",
                        base: Decimal {
                            numeric_underscore: false,
                        },
                    },
                ],
                location: 45..82,
                name: "BatchTxs",
                arguments: [
                    RecordConstructorArg {
                        decorators: [],
                        label: None,
                        annotation: Constructor {
                            location: 66..81,
                            module: None,
                            name: "List",
                            arguments: [
                                Constructor {
                                    location: 71..80,
                                    module: None,
                                    name: "ByteArray",
                                    arguments: [],
                                },
                            ],
                        },
                        location: 66..81,
                        tipo: (),
                        doc: None,
                    },
                ],
                doc: None,
                sugar: false,
            },
            RecordConstructor {
                decorators: [
                    Tag {
                        value: "1170",
                        base: Decimal {
                            numeric_underscore: false,
                        },
                    },
                ],
                location: 86..131,
                name: "Committed",
                arguments: [
                    RecordConstructorArg {
                        decorators: [],
                        label: None,
                        annotation: Constructor {
                            location: 109..113,
                            module: None,
                            name: "Bool",
                            arguments: [],
                        },
                        location: 109..113,
                        tipo: (),
                        doc: None,
                    },
                    RecordConstructorArg {
                        decorators: [],
                        label: None,
                        annotation: Constructor {
                            location: 115..130,
                            module: None,
                            name: "List",
                            arguments: [
                                Constructor {
                                    location: 120..129,
                                    module: None,
                                    name: "ByteArray",
                                    arguments: [],
                                },
                            ],
                        },
                        location: 115..130,
                        tipo: (),
                        doc: None,
                    },
                ],
                doc: None,
                sugar: false,
            },
            RecordConstructor {
                decorators: [
                    Tag {
                        value: "1171",
                        base: Decimal {
                            numeric_underscore: false,
                        },
                    },
                ],
                location: 135..183,
                name: "Withdrawn",
                arguments: [
                    RecordConstructorArg {
                        decorators: [],
                        label: None,
                        annotation: Constructor {
                            location: 158..162,
                            module: None,
                            name: "Bool",
                            arguments: [],
                        },
                        location: 158..162,
                        tipo: (),
                        doc: None,
                    },
                    RecordConstructorArg {
                        decorators: [
                            Len {
                                value: "32",
                                base: Decimal {
                                    numeric_underscore: false,
                                },
                            },
                        ],
                        label: None,
                        annotation: Constructor {
                            location: 173..182,
                            module: None,
                            name: "ByteArray",
                            arguments: [],
                        },
                        location: 164..182,
                        tipo: (),
                        doc: None,
                    },
                ],
                doc: None,
                sugar: false,
            },
        ],
        doc: None,
        location: 0..185,
        name: "L2Datum",
        opaque: false,
        parameters: [],
        public: true,
        typed_parameters: [],
    },
)
